### PR TITLE
Make VPN Network configurable

### LINKED ISSUE #8987: Make VPN Network Range Configurable
## Summary

Currently, the VPN network range used in Gardener is restricted to the hard-coded default values, `192.168.123.0/24` and `fd8f:6d53:b97a:1::/120`. We propose to make this network range configurable to offer more flexibility to both operators and users. This is achieved by a new configuration option in `gardener/gardener` and `gardener/vpn2`.

### Goals

- allow deploying shoots into infrastructure networks that use the currently hard-coded VPN network (`192.168.123.0/24`)

### Non-Goals

- lift restrictions that shoot and seed networks must not overlap with the VPN network

## Proposal

The proposal targets two projects: `gardener/gardener` and `gardener/vpn2`.

* [ ] In `gardener/gardener`, configure VPN Network on Seed Level by introducing a new optional field `Seed.spec.networks.vpn`, type `*string`
  * default: current VPN CIDR (`192.168.123.0/24`/`fd8f:6d53:b97a:1::/120`)
  * validation:
    * must be a valid CIDR matching the selected `IPFamily`
    * must have the same size as the current hard-coded network: `/24` (IPv4), `/120` (IPv6)
    * instead of validating disjointedness with the hard-coded CIDR, the seed networks must be disjoint with the specified CIDR
  * scheduling: shoot networks must be disjoint with seed VPN network (similar to other seed networks)
    * validate during admission
    * consider in scheduler during seed filtering
  * Implemented in https://github.com/gardener/gardener/pull/9597
* [x] In `gardener/vpn2`, introduce a new environment variable for configuring the VPN network to use:`VPN_NETWORK`
  * the value (if set) replaces the hard-coded default networks
  * similar to the current logic, the `/24` is split up into 4 equally-sized `/26` networks for configuring [HA VPN setups](https://github.com/gardener/gardener/issues/6890)
  * Implemented in https://github.com/gardener/vpn2/pull/78

The `gardenlet` sets the configured VPN network in the `values` provided to the `vpn-seed-server` and `vpn-shoot` `ManagedResource` deployments.

## Alternatives

* introduce a new optional field `Shoot.spec.networking.vpn`, type `*string`
  * default: current VPN CIDR (`192.168.123.0/24`/`fd8f:6d53:b97a:1::/120`)
  * validation:
    * must be a valid CIDR matching the selected `IPFamily`
    * must have the same size as the current hard-coded network: `/24` (IPv4), `/120` (IPv6)
    * `Seed`: the seed network disjointedness check with the VPN network is removed, as it can't be performed ahead of time (VPN Networks can be different for new Shoots)
  * Admission during shoot scheduling: validate that the specified VPN network is disjoint with the seed networks
  * CON: larger API surface. For the use cases we are aware of right now, the VPN CIDR range will be the same on all Shoots anyway
  * PRO: more flexibility, as VPN network can be selected individually for every `Shoot`

## References

* [GEP-14 Reversed Cluster VPN](https://github.com/gardener/gardener/blob/master/docs/proposals/14-reversed-cluster-vpn.md)
* [HA VPN Deployment](https://github.com/gardener/gardener/pull/6978)

**How to categorize this issue?**
/area networking
/kind enhancement

/cc @timebertt @Kumm-Kai

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area networking
/kind enhancement

**What this PR does / why we need it**:

This PR makes use of https://github.com/gardener/vpn2/pull/78 to allow configuring the used VPN network on the seed-level via `Seed.spec.networks.vpn`.
The field is defaulted to the current hard-coded values and non-default network CIDRs must have the same size as the current hard-coded values. Network disjointedness with other seed and shoot networks is ensured in validation, admission, and scheduling.

See the umbrella issue and the doc commit for more details on motivation and conditions.

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/gardener/issues/8987

**Special notes for your reviewer**:
Supersedes https://github.com/gardener/gardener/pull/8991

**Release note**:

```features operator
Operators can configure the network range of the shoot VPN tunnel using the new `Seed.spec.networks.vpn` field. See [this document](https://github.com/gardener/gardener/blob/master/docs/usage/reversed-vpn-tunnel.md#configurable-vpn-network) for more details.
```

### CODE REVIEW

**Comment by timebertt** [comment]
TL;DR: I pushed a new commit which should fix the e2e test failures for HA clusters.

---

Explanation:
I figured out why the e2e tests for HA clusters were consistently failing.
The gardenlet logs pointed towards kube-apiserver being unable to talk to metrics-server:
```
2024-04-16T20:38:52.370616999Z stderr F {"level":"error","ts":"2024-04-16T20:38:52.370Z","msg":"Error","controller":"shoot","object":{"name":"e2e-rotate","namespace":"garden-local"},"namespace":"garden-local","name":"e2e-rotate","reconcileID":"4a09b9c2-0273-4f05-b899-a231bb2821ff","operation":"reconcile","flow":"Shoot cluster reconciliation","task":"Labeling resources after modification of encryption config or to encrypt them with new ETCD encryption key","error":"retry failed with context deadline exceeded, last error: error discovering server preferred resources: unable to retrieve the complete list of server APIs: metrics.k8s.io/v1beta1: the server is currently unable to handle the request","stacktrace":"github.com/gardener/gardener/pkg/utils/flow.(*execution).runNode.func2\n\tgithub.com/gardener/gardener/pkg/utils/flow/flow.go:241"}
```

I could reproduce this locally:
```
$ k get apiservice v1beta1.metrics.k8s.io -oyaml
...
status:
  conditions:
  - lastTransitionTime: "2024-04-17T06:28:21Z"
    message: 'failing or missing response from https://10.3.1.18:8443/apis/metrics.k8s.io/v1beta1:
      Get "https://10.3.1.18:8443/apis/metrics.k8s.io/v1beta1": dial tcp 10.3.1.18:8443:
      connect: connection refused'
    reason: FailedDiscoveryCheck
    status: "False"
    type: Available
```

This was due to missing routes for the shoot networks in the kube-apiserver:
```
$ k -n shoot--local--e2e-rotate exec -it kube-apiserver-744ccc7b7c-cmcmc -c vpn-path-controller -- sh
~ # ip r
default via 169.254.1.1 dev eth0
10.5.0.0/26 dev tap0 proto kernel scope link src 10.5.0.8
10.5.0.64/26 dev tap1 proto kernel scope link src 10.5.0.72
10.5.0.192/26 dev bond0 proto kernel scope link src 10.5.0.214
169.254.1.1 dev eth0 scope link

$ k -n shoot--local--e2e-rotate logs kube-apiserver-744ccc7b7c-cmcmc vpn-path-controller
[Wed Apr 17 06:43:49 UTC 2024]: 194=err 195=err using
[Wed Apr 17 06:43:53 UTC 2024]: 194=err 195=err using
```

This in turn was due to the bonding device on the vpn-shoot-client side not using an IP from the configured VPN network but from the default network:
```
$ ks exec -it vpn-shoot-1 -c vpn-shoot-s1 -- sh
~ # ip a s dev bond0
6: bond0: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 56:1d:92:09:22:d6 brd ff:ff:ff:ff:ff:ff
    inet 192.168.123.194/26 scope global bond0
       valid_lft forever preferred_lft forever
    inet6 fe80::541d:92ff:fe09:22d6/64 scope link
       valid_lft forever preferred_lft forever
```

The bonding device was incorrectly configured because the `VPN_NETWORK` env var was missing in the `vpn-shoot-init` container of the `vpn-shoot`.
This is fixed in the latest commit.

File: pkg/apis/core/validation/seed_test.go [hunk 1]
```diff
@@ -473,74 +502,40 @@ var _ = Describe("Seed Validation Tests", func() {
...
-				It("should forbid Seed with overlap to default vpn range (superset)", func() {
-					// Pods CIDR overlaps with default vpn range
-					seed.Spec.Networks.Pods = "192.168.0.0/16" // 192.168.0.0 -> 192.168.255.255
-
-					errorList := ValidateSeed(seed)
-
-					Expect(errorList).To(ConsistOfFields(Fields{
-						"Type":   Equal(field.ErrorTypeInvalid),
-						"Field":  Equal("spec.networks.pods"),
-						"Detail": Equal(`must not overlap with "[]" ("192.168.123.0/24")`),
```
Comment (axel7born) [inline pkg/apis/core/validation/seed_test.go:533]: Would it make sense to keep the test for equality and superset?
Or are they useless?

File: pkg/apis/core/validation/seed.go [hunk 2]
```diff
@@ -270,6 +271,25 @@ func validateSeedNetworks(seedNetworks core.SeedNetworks, fldPath *field.Path, i
 	if seedNetworks.Nodes != nil {
 		networks = append(networks, cidrvalidation.NewCIDR(*seedNetworks.Nodes, fldPath.Child("nodes")))
 	}
+
+	if seedNetworks.VPN != nil {
+		// vpn network must be a valid network and must match the primary IP family (see below)
```
Comment (axel7born) [inline pkg/apis/core/validation/seed.go]:
```suggestion
		// VPN network must be a valid network and must match the primary IP family (see below)
```

File: pkg/apis/core/validation/seed.go [hunk 3]
```diff
@@ -270,6 +271,25 @@ func validateSeedNetworks(seedNetworks core.SeedNetworks, fldPath *field.Path, i
...
+		vpnNetwork = cidrvalidation.NewCIDR(*seedNetworks.VPN, fldPath.Child("vpn"))
+		// vpn network must be disjoint with the other networks (see below)
```
Comment (axel7born) [inline pkg/apis/core/validation/seed.go]:
```suggestion
		// VPN network must be disjoint with the other networks (see below)
```

File: pkg/apis/core/validation/seed.go [hunk 4]
```diff
@@ -270,6 +271,25 @@ func validateSeedNetworks(seedNetworks core.SeedNetworks, fldPath *field.Path, i
...
+		networks = append(networks, vpnNetwork)
+
+		// vpn network must be a /24 network for IPv4 and /120 for IPv6
```
Comment (axel7born) [inline pkg/apis/core/validation/seed.go]:
```suggestion
		// VPN network must be a /24 network for IPv4 and /120 for IPv6
```

File: pkg/apis/core/validation/seed_test.go [hunk 5]
```diff
@@ -473,74 +502,40 @@ var _ = Describe("Seed Validation Tests", func() {
...
-					shootDefaultPodCIDR := "192.168.123.128/28"     // 192.168.123.128 -> 192.168.123.143
-					shootDefaultServiceCIDR := "192.168.123.200/32" // 192.168.123.200 -> 192.168.123.200
-
-					nodesCIDR := "192.168.123.0/27" // 192.168.123.0 -> 192.168.123.31
-					// Nodes network overlaps with default vpn range
-					// Pods CIDR overlaps with default vpn range
-					// Services CIDR overlaps with default vpn range
-					// Shoot default pod CIDR overlaps with default vpn range
-					// Shoot default service CIDR overlaps with default vpn range
+				It("should forbid Seed with overlap to vpn range", func() {
```
Comment (axel7born) [inline pkg/apis/core/validation/seed_test.go]:
```suggestion
				It("should forbid Seed with overlap to VPN range", func() {
```

File: pkg/gardenlet/operation/botanist/vpnshoot.go [hunk 6]
```diff
@@ -32,6 +34,7 @@ func (b *Botanist) DefaultVPNShoot() (vpnshoot.Interface, error) {
 		Image:      image.String(),
 		VPAEnabled: b.Shoot.WantsVerticalPodAutoscaler,
 		ReversedVPN: vpnshoot.ReversedVPNValues{
+			VPNCIDR:     ptr.Deref(b.Seed.GetInfo().Spec.Networks.VPN, ""),
```
Comment (axel7born) [inline pkg/gardenlet/operation/botanist/vpnshoot.go]: The empty string as default looks confusing to me.
Can't we just derefence?

File: pkg/scheduler/controller/shoot/reconciler_test.go [hunk 7]
```diff
@@ -682,6 +683,25 @@ var _ = Describe("Scheduler_Control", func() {
 			Expect(bestSeed).To(BeNil())
 		})
 
+		It("should fail because it cannot find a seed cluster due to network overlap with vpn network", func() {
```
Comment (axel7born) [inline pkg/scheduler/controller/shoot/reconciler_test.go]:
```suggestion
		It("should fail because it cannot find a seed cluster due to network overlap with VPN network", func() {
```

File: pkg/utils/validation/cidr/disjoint.go [hunk 8]
```diff
@@ -49,12 +47,8 @@ func validateOverlapWithSeed(fldPath *field.Path, shootNetwork *string, networkT
...
 		}
 
-		if NetworksIntersect(v1beta1constants.DefaultVPNRange, *shootNetwork) {
-			allErrs = append(allErrs, field.Invalid(fldPath, *shootNetwork, fmt.Sprintf("shoot %s network intersects with default vpn network (%s)", networkType, v1beta1constants.DefaultVPNRange)))
-		}
-
-		if NetworksIntersect(v1beta1constants.DefaultVPNRangeV6, *shootNetwork) {
-			allErrs = append(allErrs, field.Invalid(fldPath, *shootNetwork, fmt.Sprintf("shoot %s network intersects with default vpn network (%s)", networkType, v1beta1constants.DefaultVPNRangeV6)))
+		if seedVPN != nil && NetworksIntersect(*seedVPN, *shootNetwork) {
+			allErrs = append(allErrs, field.Invalid(fldPath, *shootNetwork, fmt.Sprintf("shoot %s network intersects with seed vpn network", networkType)))
```
Comment (axel7born) [inline pkg/utils/validation/cidr/disjoint.go]: VPN in caps? This also needs to be adapted in the tests.
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath, *shootNetwork, fmt.Sprintf("shoot %s network intersects with seed VPN network", networkType)))
```

**Comment by axel7born** [comment]
Thank you for your change.
To me it looks good.
I'm not so sure about the `vpn` vs `VPN`.

File: pkg/apis/core/validation/seed_test.go (hunk 1 above)
Comment (timebertt) [reply pkg/apis/core/validation/seed_test.go:533]: I combined the three test cases into `should forbid Seed with overlap to vpn range`. The different seed networks are a subset/superset of and/or equal to the VPN network.
I think all cases are covered by this :)

File: pkg/gardenlet/operation/botanist/vpnshoot.go (hunk 6 above)
Comment (timebertt) [reply pkg/gardenlet/operation/botanist/vpnshoot.go]: You're right, we also do this for optional defaulted fields in other places. Changed to deref.

**Comment by timebertt** [comment]
Thanks @axel7born for your review. I addressed the suggestions and capitalized `VPN` consistently across the entire codebase.
PTAL :)

File: docs/usage/reversed-vpn-tunnel.md [hunk 9]
```diff
@@ -19,6 +19,22 @@ Connection establishment with a reversed tunnel:
...
+
+Similar to the other seed networks (see `Seed.spec.networks`), the configured VPN network must not overlap with any other seed network.
+Also, shoot networks (see `Shoot.spec.networking`) must not overlap with any seed network including the VPN network.
+This is required for packets to be routed without ambiguity in all components of the seed/shoot architecture.
+Migrating the control plane of a shoot to a seed cluster with a different VPN network configuration works as long as network disjointedness is still fulfilled.
+
+With the requirement for disjoint seed and shoot networks, Gardener users are limited in choosing network ranges for their shoots.
+Gardener operators might want to use coherent ranges for their seed networks including the VPN network to give users more freedom of choice and make the limitations simpler to reason about.
+For this, configuring a non-default VPN network on the seed-level allows Gardener operators to use for example the [IANA-Reserved IPv4 Shared Address Space](https://datatracker.ietf.org/doc/html/rfc6598#section-7) (`100.64.0.0/10`) for all seed networks (the ones that users can't choose for their shoots).
+
```
Comment (ScheererJ) [inline docs/usage/reversed-vpn-tunnel.md:37]: Overall, the motivation to make the VPN range configurable seems to come from the desire to make as much [private IP space](https://datatracker.ietf.org/doc/html/rfc1918) as possible available to shoot owners. As the network used here is not routed within the seed/shoot, i.e. there are no packets on the wire with for example `192.168.123.1` as IP addresses, would it not be a better solution to simply switch the VPN to a pure IPv6 tunnel with potentially IPv4 or IPv6 workload to be transferred? Did you consider this approach?

File: pkg/apis/core/v1beta1/types_seed.go [hunk 10]
```diff
@@ -196,6 +196,13 @@ type SeedNetworks struct {
 	Pods string `json:"pods" protobuf:"bytes,2,opt,name=pods"`
 	// Services is the CIDR of the service network. This field is immutable.
 	Services string `json:"services" protobuf:"bytes,3,opt,name=services"`
+	// VPN is the CIDR of the vpn network (spanning vpn-seed-server and vpn-shoot) of shoots on this seed. It must be
+	// disjoint with the Nodes, Pods, and Services networks. It must be a /24 network for IPv4 and /120 for IPv6.
+	// This field is mutable. Changing it leads to a temporary VPN disconnect during the next reconciliation of all shoots
+	// on this seed.
+	// Defaults to "192.168.123.0/24" or "fd8f:6d53:b97a:1::/120" depending on the primary IPFamilies configuration.
+	// +optional
+	VPN *string `json:"vpn,omitempty" protobuf:"bytes,7,opt,name=vpn"`
```
Comment (ScheererJ) [inline pkg/apis/core/v1beta1/types_seed.go]: Why did you add the new field in the middle of the struct? It is confusing and may lead to follow-up errors as the order of the fields in `protobuf` does not match to the field order.

File: pkg/apis/core/validation/seed.go [hunk 11]
```diff
@@ -270,6 +271,25 @@ func validateSeedNetworks(seedNetworks core.SeedNetworks, fldPath *field.Path, i
...
+		if vpnCIDR := vpnNetwork.GetIPNet(); vpnCIDR != nil {
+			prefix, _ := vpnCIDR.Mask.Size()
+			if primaryIPFamily == core.IPFamilyIPv4 && prefix != 24 {
+				allErrs = append(allErrs, field.Invalid(vpnNetwork.GetFieldPath(), vpnNetwork.GetCIDR(), "VPN network must have a /24 prefix for IPv4 seeds"))
+			}
+			if primaryIPFamily == core.IPFamilyIPv6 && prefix != 120 {
+				allErrs = append(allErrs, field.Invalid(vpnNetwork.GetFieldPath(), vpnNetwork.GetCIDR(), "VPN network must have a /120 prefix for IPv6 seeds"))
+			}
```
Comment (ScheererJ) [inline pkg/apis/core/validation/seed.go:279]: Not sure if we need check for `/24` and `/120`, but we could just check if the network is at least 8-bit in size, e.g. via

```
if prefix, total := vpnCIDR.Mask.Size(); total - prefix < 8 {
	allErrs = append(...)
}
```

The error message might not be as direct because there is no protocol distinction, but we do not explicitly deal with `/24` and `/120`. Instead we focus on the important aspect, i.e. having at least 256 addresses available in the range.

WDYT?

File: plugin/pkg/shoot/validator/admission.go [hunk 12]
```diff
@@ -747,6 +747,7 @@ func (c *validationContext) validateShootNetworks(a admission.Attributes, worker
 				c.shoot.Spec.Networking.Pods,
 				c.shoot.Spec.Networking.Services,
 				c.seed.Spec.Networks.Nodes,
+				c.seed.Spec.Networks.VPN,
```
Comment (ScheererJ) [inline plugin/pkg/shoot/validator/admission.go:739]: This validation only happens when the control plane is migrated. Where is the validation logic (if any) that prevents operations from accidentally changing the VPN range so that it overlaps with an existing shoot range hosted on the seed?

File: pkg/apis/core/v1beta1/constants/types_constants.go [hunk 13]
```diff
@@ -749,11 +749,6 @@ const (
 	// or the specified namespace was not present.
 	NamespaceCreatedByProjectController = "namespace.gardener.cloud/created-by-project-controller"
 
-	// DefaultVPNRange is the default IPv4 network range for the VPN between seed and shoot cluster.
-	DefaultVPNRange = "192.168.123.0/24"
-	// DefaultVPNRangeV6 is the default IPv6 network range for the VPN between seed and shoot cluster.
-	DefaultVPNRangeV6 = "fd8f:6d53:b97a:1::/120"
-
```
Comment (ScheererJ) [inline pkg/apis/core/v1beta1/constants/types_constants.go:787]: Could you please share the reasoning why you moved the constants? It looks like you moved the constants from the `constants` package one layer above. The values are still constants, though.

File: pkg/gardenlet/operation/botanist/kubeapiserver.go [hunk 14]
```diff
@@ -59,8 +59,9 @@ func (b *Botanist) DefaultKubeAPIServer(ctx context.Context) (kubeapiserver.Inte
 
 	if !b.Shoot.IsWorkerless {
 		vpnConfig.Enabled = true
+		vpnConfig.VPNNetworkCIDR = ptr.Deref(b.Seed.GetInfo().Spec.Networks.VPN, "")
```
Comment (ScheererJ) [inline pkg/gardenlet/operation/botanist/kubeapiserver.go]: Just wondering: Is there a valid point in time when `spec.networks.vpn` should not be set and this code executes? It seems like dereferencing should always work or am I missing something.

How does the new vpn code handle an empty `VPN_NETWORK` environment variable?

File: pkg/gardenlet/operation/botanist/vpnseedserver.go [hunk 15]
```diff
@@ -41,9 +41,10 @@ func (b *Botanist) DefaultVPNSeedServer() (vpnseedserver.Interface, error) {
 		ImageAPIServerProxy:      imageAPIServerProxy.String(),
 		ImageVPNSeedServer:       imageVPNSeedServer.String(),
 		Network: vpnseedserver.NetworkValues{
+			VPNCIDR:     ptr.Deref(b.Seed.GetInfo().Spec.Networks.VPN, ""),
```
Comment (ScheererJ) [inline pkg/gardenlet/operation/botanist/vpnseedserver.go]: Same question as in `kubeapiserver.go` with regards to `spec.networks.vpn`.

File: pkg/gardenlet/operation/botanist/vpnseedserver.go [hunk 16]
```diff
@@ -41,9 +41,10 @@ func (b *Botanist) DefaultVPNSeedServer() (vpnseedserver.Interface, error) {
...
 			PodCIDR:     b.Shoot.Networks.Pods.String(),
 			ServiceCIDR: b.Shoot.Networks.Services.String(),
-			// NodeCIDR is set in DeployVPNServer to handle dynamice node network CIDRs
+			// NodeCIDR is set in DeployVPNServer to handle dynamic node network CIDRs
```
Comment (ScheererJ) [inline pkg/gardenlet/operation/botanist/vpnseedserver.go:37]: Nice consistency in the typos across file boundaries ;-) ...

File: example/gardener-local/gardenlet/values.yaml [hunk 17]
```diff
@@ -92,6 +92,7 @@ config:
         # Those CIDRs must match those specified in the kind Cluster configuration.
         pods: 10.1.0.0/16
         services: 10.2.0.0/16
+        vpn: 10.5.0.0/24
```
Comment (ScheererJ) [inline example/gardener-local/gardenlet/values.yaml]: While I can understand the desire to test the feature in the end-to-end tests it also means that the default case is no longer tested in them. Can we find a middle ground?

File: pkg/component/networking/vpn/shoot/shoot.go [hunk 18]
```diff
@@ -614,6 +607,14 @@ func (v *vpnShoot) getEnvVars(index *int) []corev1.EnvVar {
 	return envVariables
 }
 
+func (v *vpnShoot) getIPFamiliesEnvVar() string {
```
Comment (ScheererJ) [inline pkg/component/networking/vpn/shoot/shoot.go]:
```suggestion
func (v *vpnShoot) getIPFamiliesEnvVarValue() string {
```

**Review by ScheererJ — CHANGES_REQUESTED** [review]
Thanks a lot for the enhancement. I have a few questions/comments.

File: docs/usage/reversed-vpn-tunnel.md (hunk 9 above)
Comment (timebertt) [reply docs/usage/reversed-vpn-tunnel.md:37]: > the motivation to make the VPN range configurable seems to come from the desire to make as much [private IP space](https://datatracker.ietf.org/doc/html/rfc1918) as possible available to shoot owners

Correct.

> As the network used here is not routed within the seed/shoot, i.e. there are no packets on the wire with for example `192.168.123.1` as IP addresses, would it not be a better solution to simply switch the VPN to a pure IPv6 tunnel with potentially IPv4 or IPv6 workload to be transferred?

We didn't consider this approach. Maybe we could have considered it if it had been suggested in the proposal issue 😄 

Even if there are no packets on the wire with addresses from the VPN network, I assume that there must not be overlaps between seed/shoot networks and the VPN network. Otherwise, the routes will be ambiguous. Do I get this right?
If so, using an IPv6 VPN network might solve the problem for IPv4 shoots (obviously). Nevertheless, choosing a hard-coded IPv6 network will yield the same "problem" that it imposes additional restrictions (apart from seed networks) on what networks can be chosen for shoots. I quote "problem" because it will be less restrictive in IPv6.
I still think operators would like to put all "restricted" networks (seeds and VPN) into the same IP space, even for IPv6.
Feel free to correct me if my naive thinking doesn't make sense 😄 

That being said, I would love to continue with the current approach as it's almost complete. I feel like it doesn't bring too much complexity that it's worth going back to the drawing board and implementing a totally different approach.

Comment (axel7born) [reply docs/usage/reversed-vpn-tunnel.md:37]: Please continue with this approach!
Johannes is on vacation this week. But we had a conversation on Friday, where he said this idea should not block this PR.

File: pkg/apis/core/v1beta1/types_seed.go (hunk 10 above)
Comment (timebertt) [reply pkg/apis/core/v1beta1/types_seed.go]: FMPOV, it made sense to add the new field after the other network configuration options.

The protobuf generator can handle struct fields in a different order than in the protobuf spec.
When we remove fields, we have to leave tombstones for the protobuf generator, but this applies to the general case, not only here. See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md:

> [...] if the feature is abandoned, or the field name is changed, the field should be removed from the go struct, with a tombstone comment ensuring the field name and protobuf tag are not reused:

There are even k8s core API types that have struct fields that are out-of-order in terms of their protobuf fields: [`PodSpec`](https://github.com/kubernetes/api/blob/039b247b85af57bb7144c894424602b57dd1e98d/core/v1/types.go#L3601), [`Secret`](https://github.com/kubernetes/api/blob/039b247b85af57bb7144c894424602b57dd1e98d/core/v1/types.go#L6970). So we should be safe here.

That being said, I'm willing to change the order if you insist :)

File: pkg/apis/core/validation/seed.go (hunk 11 above)
Comment (timebertt) [reply pkg/apis/core/validation/seed.go:279]: You're right, we could have implemented it this way. However, in https://github.com/gardener/vpn2/pull/78 kept it simple and ensures that `VPN_NETWORK` is set to a `/24` or `/120` net: https://github.com/gardener/vpn2/blob/08a3f75ad01e9e249d7319eeb984b5b2125aa05d/seed-server/network-connection.sh#L50-L53

We can still lift this restriction if there is a need. But it will require some more bash magic in vpn2 😄

File: plugin/pkg/shoot/validator/admission.go (hunk 12 above)
Comment (timebertt) [reply plugin/pkg/shoot/validator/admission.go:739]: Nice catch! You're right, cross-validation for changing `Seed.spec.networks.vpn` is missing. I will add it.

File: pkg/apis/core/v1beta1/constants/types_constants.go (hunk 13 above)
Comment (timebertt) [reply pkg/apis/core/v1beta1/constants/types_constants.go:787]: I wanted to keep it similar to other default values, which are also constants in the versioned API package itself. In a new version of the API, we might apply different defaults, so it's probably sensible to put it into the versioned package.
I don't have a strong opinion on where to put them though, because both packages are used for constants in different places.

File: pkg/gardenlet/operation/botanist/kubeapiserver.go (hunk 14 above)
Comment (timebertt) [reply pkg/gardenlet/operation/botanist/kubeapiserver.go]: Yep, already changed this to direct dereferencing in 2eacc83560c71415c8b0c85678e252b74616133b as suggested in https://github.com/gardener/gardener/pull/9597#discussion_r1570903823

File: pkg/gardenlet/operation/botanist/vpnseedserver.go (hunk 15 above)
Comment (timebertt) [reply pkg/gardenlet/operation/botanist/vpnseedserver.go]: See above

File: example/gardener-local/gardenlet/values.yaml (hunk 17 above)
Comment (timebertt) [reply example/gardener-local/gardenlet/values.yaml]: Good point. To verify both the default case and a non-default configuration, I changed the local seed back to the default configuration and changed the local2 seed to the non-default configuration.
With this, both cases are verified in e2e tests during control plane migrations (modulo https://github.com/gardener/gardener/issues/9604).

**Comment by timebertt** [comment]
@axel7born I addressed all suggestions added by @ScheererJ. Can you take another look?

**Comment by timebertt** [comment]
Rebased and re-generated `pkg/apis/core/v1beta1/generated.pb.go` to resolve conflicts.

**Comment by timebertt** [comment]
@ScheererJ can you take another look? :)

File: docs/usage/reversed-vpn-tunnel.md (hunk 9 above)
Comment (ScheererJ) [reply docs/usage/reversed-vpn-tunnel.md:37]: > Maybe we could have considered it if it had been suggested in the proposal issue 😄

As you know, feedback can come at any point in time. An idea that does not fit to your timeline is not automatically a bad one.

I disagree with your opinion that with IPv6 it would be a problem. If we would use the hard-coded ULA range as done now with IPv6-only clusters we would only have a very unlikely collision if someone else used the same range. ([RFC 4193](https://datatracker.ietf.org/doc/html/rfc4193#section-3.2) requires the global id to be random.)
It would even be possible to reserve a separate global range or use some of the special IPv6 ranges to prevent any chance of collision at all. Therefore, I do believe that you can solve the problem of the scarce IPv4 private address space by using IPv6 for the VPN.
Feel free to convince me otherwise, though.

The overlap concerns between seed and shoot clusters with regards to node, pods and services have not a lot to do with the choice of the VPN range. In other words, the seed/shoot disjointness still needs to be ensured regardless what happens with the VPN.

As the VPN range is not routed in the network, I am not sure if operators are that much interested in it. Of course, if it causes other restrictions it may be relevant, but that is again only relevant in the context of scarce IPv4 space.

Feel free to continue with the approach. As @axel7born already mentioned, I have no hard feelings against this approach, but I do feel that we just use another set of band-aid instead of actually solving the problem.

File: pkg/apis/core/v1beta1/types_seed.go (hunk 10 above)
Comment (ScheererJ) [reply pkg/apis/core/v1beta1/types_seed.go]: Please change the order.

I was not concerned about the protobuf generator at all, but only about humans reading/changing the code. As far as I can tell we append only, which makes this a rare exception that can be easily overlooked. I would like to prevent future (human) errors by simply sticking to the append-only approach.

File: pkg/apis/core/v1beta1/constants/types_constants.go (hunk 13 above)
Comment (ScheererJ) [reply pkg/apis/core/v1beta1/constants/types_constants.go:787]: Fine with me. I doubt that these ranges change a lot over time, though.

File: plugin/pkg/utils/miscellaneous.go [hunk 19]
```diff
@@ -88,3 +91,35 @@ func ValidateZoneRemovalFromSeeds(oldSeedSpec, newSeedSpec *core.SeedSpec, seedN
...
+			shoot.Spec.Networking.Services,
+			newSeedSpec.Networks.Nodes,
+			newSeedSpec.Networks.VPN,
+			newSeedSpec.Networks.Pods,
+			newSeedSpec.Networks.Services,
+			v1beta1helper.IsWorkerless(shoot),
+		)) > 0 {
+			return apierrors.NewForbidden(core.Resource(kind), seedName, fmt.Errorf("cannot update networks of %s %s as there are Shoots with overlapping networks scheduled to this Seed", kind, seedName))
+		}
+	}
```
Comment (ScheererJ) [inline plugin/pkg/utils/miscellaneous.go]: Why not add more context to the error message? The validation actually checks, which shoot has an overlap in what range, but we simply discard this knowledge and only return a generic message.

File: docs/usage/reversed-vpn-tunnel.md (hunk 9 above)
Comment (timebertt) [reply docs/usage/reversed-vpn-tunnel.md:37]: > An idea that does not fit to your timeline is not automatically a bad one.

Definitely not! I didn't want to say so.

> I do feel that we just use another set of band-aid instead of actually solving the problem.

Agreed.

TBH, I do find your suggestion compelling but I lack some knowledge to comprehend it fully.
I would like to discuss the details of your idea. A meeting will probably be difficult this week, so I propose taking this to next week's hackathon :)

In the meantime, I will address the remaining comments in this PR so that we have a finished implementation to go back to if we decide on this approach after all.

**Comment by timebertt** [comment]
/hold
We want to discuss the overall approach for solving the motivating problem, see https://github.com/gardener/gardener/pull/9597#discussion_r1590576312

File: pkg/apis/core/v1beta1/types_seed.go (hunk 10 above)
Comment (timebertt) [reply pkg/apis/core/v1beta1/types_seed.go]: Got it, your reasoning makes sense to me. I moved the new fields to the end of the structs.

File: plugin/pkg/utils/miscellaneous.go (hunk 19 above)
Comment (timebertt) [reply plugin/pkg/utils/miscellaneous.go]: Done. I hope this is what you had in mind.

**Comment by rfranzke** [comment]
Shall we close this PR in favor of https://github.com/gardener-community/hackathon/blob/main/2024-05_Schelklingen/README.md#%EF%B8%8F-pure-ipv6-based-vpn-tunnel? :)
